docs/adr/0023-paypal-support.md

```diff
@@ -0,0 +1,102 @@
+
+## Decision Drivers
+
+- Engineering resources
+- Future effort needed to integrate IAP
+- Code complexity as it relates to existing Subscription Platform infrastructure
+
+## Considered Options
+
+- Stripe-driven out-of-band invoice processing
+- PayPal subscriptions linked in FxA
+- Recurly
+
+## Decision Outcome
+
+Originally considered to be the 'fast-path', using PayPal subscriptions ended up being approximately the same engineering effort to implement as Stripe-driven invoice processing. Given additional user experience benefits of the Stripe-driven approach, it is now the recommend implementation path.
+
+## Pros and Cons of the Options
+
+### Stripe-driven out-of-band invoice processing
+
+Stripe defaults subscriptions to billing automatically using a payment source on file with Stripe. Stripe also provides the option to have subscriptions that have a `collection_method` of `send_invoice`. In this model, Stripe can track subscriptions, and generate invoices that can be paid `out-of-band`. This allows Stripe to have an accurate record of charges, invoices, and subscriptions but how the invoice is considered 'paid' is up to FxA.
+
+PayPal has a feature called reference transactions that allow FxA to get a customers authorization to charge their PayPal account. This authorization allows for later payments to be charged against the PayPal account. We will track this reference ID in Stripe metadata for the customer and mark the subscription as `send_invoice`.
+
+FxA will implement additional logic to listen for invoice creation webhooks and switch their `auto_advance` off if the subscription has metadata indicating it should be paid via a PayPal reference ID. We will then check the invoice for the amount to charge and run the PayPal charge via PayPal API. The invoice will then be updated to reflect whether it was paid successfully or not.
+
+Pros:
+
+- Stripe continues to serve as an accurate source of truth regarding a customers subscription.
+- All the logic that determines active subscriptions and entitlements using Stripe can be used as-is.
+- Multiple product support works as-is.
+- Support representatives can see accurate payment history in Stripe for customers paying via PayPal.
+- Support can credit a PayPal customer, just as they can for a credit card customer.
+- Can extend in a similar manner to track a correlating IAP for a user.
+- Can handle switching payment method from credit card to PayPal reference ID.
+
+Cons:
+
+- Additional metadata in Stripe can start pushing us closer to metadata limits on objects.
+- Sensitive processing logic that runs PayPal transactions which we need to ensure work correctly.
+- We still could end up storing PayPal reference ID's in a separate db table to ensure we track
+  prior PayPal billing.
```

Comment:
A reference transaction ID is not per subscription, its per customer as its a blanket permission to charge the PayPal account. If a user cancels a subscription, we will not cancel the reference transaction ID until the subscription periods have expired and there are no other subscriptions active. The table here is mainly for history, so that we know the user previously paid with PayPal, and can pull the cancelled billing agreement up with the id.

Reply:
My understanding is that as long as the user is not cancelling the billing agreement as mentioned [here](https://developer.paypal.com/docs/archive/express-checkout/ec-set-up-reference-transactions/#cancel-a-billing-agreement) I think we can continue to use the reference ID to resubscribe or even create new subscriptions. I found this document helpful in understanding the relationship between the reference ID and the billing agreement. 

I am thinking in the DB we will only need the following columns: a PayPal reference ID, Stripe customer ID, and FXA user ID. Since having these information alone will allow us to get everything we need to validate, cancel, or create new subscriptions.

Also new to ADRs, but I think the specifics on columns for the table can be further narrowed after a decision is made?
@bbangert, can you share your thoughts here?

Reply:
ADRs are for overall abstract architectural concerns. If an exact DB table schema doesn't swing the decision one way or another, it doesn't really need to be nailed down in the ADR. 

But, unless I'm mistaken, I think the concern here is whether this weakens the notion that Stripe is a single source of subscription truth - because we'd store this payment-related data in more than one place (i.e. Stripe customer metadata *and/or* FxA DB table)? Not sure whether that's a huge concern, but it's worth noting.